File: java/numberOfSubstringsContainingAllThreeCharacters.java
class Solution {
    public int numberOfSubstrings(String s) {
        int totalSubstrings = 0;
        //map will keep track of how many of each character is currently in each substring
        Map<Character, Integer> numChars = new HashMap<>();
        numChars.put('a', 0);
        numChars.put('b', 0);
        numChars.put('c', 0);
        int substringEnd = -1;
        int substringStart = 0;
        while (substringEnd < s.length()) {
            //increment substringEnd until the substring contains all three characters
            while (!containsAbc(numChars) && substringEnd < s.length()) {
                substringEnd++;
                if (substringEnd < s.length()) {
                    numChars.put(s.charAt(substringEnd), numChars.get(s.charAt(substringEnd)) + 1);
                }
            }
            //if there are no possible substrings, break from the loop
            if (substringEnd >= s.length()) {
                break;
            }
            //increment the total by all substrings with the given starting index and an ending index equal to or greater than the current end index
            totalSubstrings += s.length() - substringEnd;
            //increment the starting index and remove the first character of the substring from the set
            numChars.put(s.charAt(substringStart), numChars.get(s.charAt(substringStart)) - 1);
            substringStart++;
        }
        return totalSubstrings;
    }
        
    public boolean containsAbc(Map numChars) {
        if ((int) numChars.get('a') == 0 || (int) numChars.get('b') == 0 || (int) numChars.get('c') == 0) {
            return false;
        }
        return true;
    }
}

File: python/number_of_substrings_containing_all_three_characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        #keep track of how many of each character is in each substring
        num_chars = { "a": 0, "b": 0, "c": 0 }
        start_index, end_index = 0, -1
        total_substrings = 0
        while end_index < len(s):
            #increment end_index until substring is valid
            while not self.containsAbc(num_chars) and end_index < len(s):
                end_index += 1
                if end_index < len(s):
                    num_chars[s[end_index]] += 1
            #if there are no valid substrings for the given start index, exit the while loop
            if end_index >= len(s):
                break
            #add the number of substrings with the given start index and end index greater or equal to end_index
            total_substrings += len(s) - end_index
            num_chars[s[start_index]] -= 1
            start_index += 1
        return total_substrings
        
    def containsAbc(self, num_chars: dict) -> bool:
        if num_chars["a"] == 0 or num_chars["b"] == 0 or num_chars["c"] == 0:
            return False
        return True